File: nodejs-bot/src/exchange.js
/**
 * Binance Exchange API Wrapper
 * ============================
 * Handles all communication with Binance API
 */

const Binance = require('binance-api-node').default;

class Exchange {
    constructor(apiKey, secretKey, paperMode = true) {
        this.paperMode = paperMode;
        this.paperBalance = 300; // Starting paper balance in USDT
        this.paperPositions = [];
        
        // Initialize Binance client
        if (!paperMode && apiKey && secretKey) {
            this.client = Binance({
                apiKey: apiKey,
                apiSecret: secretKey,
            });
        } else {
            // Public client for market data
            this.client = Binance();
        }
        
        this.priceCache = {};
        this.lastPriceFetch = 0;
    }
    
    /**
     * Test API connection
     */
    async testConnection() {
        try {
            await this.client.time();
            return true;
        } catch (error) {
            console.error('Connection test failed:', error.message);
            return false;
        }
    }
    
    /**
     * Get account balance in USDT (including all assets converted to USDT value)
     */
    async getBalance() {
        if (this.paperMode) {
            return this.paperBalance;
        }
        
        try {
            const account = await this.client.accountInfo();
            const usdtBalance = account.balances.find(b => b.asset === 'USDT');
            return parseFloat(usdtBalance?.free || 0);
        } catch (error) {
            console.error('Failed to get balance:', error.message);
            return 0;
        }
    }
    
    /**
     * Get TOTAL portfolio value (all assets converted to USDT)
     */
    async getTotalBalanceUSDT() {
        if (this.paperMode) {
            return this.paperBalance;
        }
        
        try {
            const account = await this.client.accountInfo();
            const allPrices = await this.client.prices();
            
            let totalUSDT = 0;
            const holdings = [];
            
            for (const balance of account.balances) {
                const free = parseFloat(balance.free);
                const locked = parseFloat(balance.locked);
                const total = free + locked;
                
                if (total > 0) {
                    let valueUSDT = 0;
                    
                    if (balance.asset === 'USDT') {
                        valueUSDT = total;
                    } else if (balance.asset === 'BUSD' || balance.asset === 'USDC') {
                        valueUSDT = total; // Stablecoins ~1:1
                    } else {
                        // Try to get USDT price
                        const usdtPair = `${balance.asset}USDT`;
                        if (allPrices[usdtPair]) {
                            valueUSDT = total * parseFloat(allPrices[usdtPair]);
                        }
                    }
                    
                    if (valueUSDT > 0.01) {
                        totalUSDT += valueUSDT;
                        holdings.push({
                            asset: balance.asset,
                            amount: total,
                            free: free,
                            locked: locked,
                            valueUSDT: valueUSDT
                        });
                    }
                }
            }
            
            console.log(`\n📊 Portfolio Holdings:`);
            holdings.forEach(h => {
                console.log(`   ${h.asset}: ${h.amount.toFixed(6)} (~$${h.valueUSDT.toFixed(2)} USDT)`);
            });
            console.log(`   💰 TOTAL: $${totalUSDT.toFixed(2)} USDT\n`);
            
            return { totalUSDT, holdings, freeUSDT: parseFloat(account.balances.find(b => b.asset === 'USDT')?.free || 0) };
        } catch (error) {
            console.error('Failed to get total balance:', error.message);
            return { totalUSDT: 0, holdings: [], freeUSDT: 0 };
        }
    }
    
    /**
     * Get available USDT for trading (free balance only)
     */
    async getAvailableUSDT() {
        if (this.paperMode) {
            return this.paperBalance;
        }
        
        try {
            const account = await this.client.accountInfo();
            const usdtBalance = account.balances.find(b => b.asset === 'USDT');
            return parseFloat(usdtBalance?.free || 0);
        } catch (error) {
            console.error('Failed to get available USDT:', error.message);
            return 0;
        }
    }
    
    /**
     * Get current price for a symbol
     */
    async getPrice(symbol) {
        try {
            const ticker = await this.client.prices({ symbol });
            const price = parseFloat(ticker[symbol]);
            this.priceCache[symbol] = price;
            return price;
        } catch (error) {
            console.error(`Failed to get price for ${symbol}:`, error.message);
            return this.priceCache[symbol] || 0;
        }
    }
    
    /**
     * Get prices for multiple symbols
     */
    async getPrices(symbols) {
        try {
            const allPrices = await this.client.prices();
            const result = {};
            symbols.forEach(symbol => {
                result[symbol] = parseFloat(allPrices[symbol] || 0);
                this.priceCache[symbol] = result[symbol];
            });
            return result;
        } catch (error) {
            console.error('Failed to get prices:', error.message);
            return symbols.reduce((acc, s) => ({ ...acc, [s]: this.priceCache[s] || 0 }), {});
        }
    }
    
    /**
     * Get historical candles (OHLCV data)
     */
    async getCandles(symbol, interval = '15m', limit = 100) {
        try {
            const candles = await this.client.candles({
                symbol,
                interval,
                limit
            });
            
            return candles.map(c => ({
                timestamp: c.openTime,
                open: parseFloat(c.open),
                high: parseFloat(c.high),
                low: parseFloat(c.low),
                close: parseFloat(c.close),
                volume: parseFloat(c.volume),
            }));
        } catch (error) {
            console.error(`Failed to get candles for ${symbol}:`, error.message);
            return [];
        }
    }
    
    /**
     * Get symbol info (minimum quantities, price precision, etc.)
     */
    async getSymbolInfo(symbol) {
        try {
            const info = await this.client.exchangeInfo();
            const symbolInfo = info.symbols.find(s => s.symbol === symbol);
            
            if (!symbolInfo) return null;
            
            const lotSizeFilter = symbolInfo.filters.find(f => f.filterType === 'LOT_SIZE');
            const priceFilter = symbolInfo.filters.find(f => f.filterType === 'PRICE_FILTER');
            const minNotionalFilter = symbolInfo.filters.find(f => f.filterType === 'NOTIONAL' || f.filterType === 'MIN_NOTIONAL');
            
            return {
                symbol,
                baseAsset: symbolInfo.baseAsset,
                quoteAsset: symbolInfo.quoteAsset,
                minQty: parseFloat(lotSizeFilter?.minQty || 0),
                maxQty: parseFloat(lotSizeFilter?.maxQty || 0),
                stepSize: parseFloat(lotSizeFilter?.stepSize || 0),
                tickSize: parseFloat(priceFilter?.tickSize || 0),
                minNotional: parseFloat(minNotionalFilter?.minNotional || minNotionalFilter?.notional || 10),
            };
        } catch (error) {
            console.error(`Failed to get symbol info for ${symbol}:`, error.message);
            return null;
        }
    }
    
    /**
     * Round quantity to valid step size
     */
    roundQuantity(quantity, stepSize) {
        const precision = Math.max(0, Math.round(-Math.log10(stepSize)));
        return Math.floor(quantity * Math.pow(10, precision)) / Math.pow(10, precision);
    }
    
    /**
     * Get valid quantity for a symbol (rounds to step size and checks min)
     */
    async getValidQuantity(symbol, quantity) {
        const info = await this.getSymbolInfo(symbol);
        if (!info) return quantity;
        
        // Round to step size
        let validQty = this.roundQuantity(quantity, info.stepSize);
        
        // Ensure minimum
        if (validQty < info.minQty) {
            return 0;
        }
        
        return validQty;
    }
    
    /**
     * Execute a market buy order
     */
    async marketBuy(symbol, quantity) {
        const price = await this.getPrice(symbol);
        
        // Get valid quantity for this symbol
        const validQty = await this.getValidQuantity(symbol, quantity);
        if (validQty <= 0) {
            throw new Error(`Quantity ${quantity} too small for ${symbol} (min lot size not met)`);
        }
        
        if (this.paperMode) {
            const cost = validQty * price;
            if (cost > this.paperBalance) {
                throw new Error('Insufficient paper balance');
            }
            
            this.paperBalance -= cost;
            this.paperPositions.push({
                symbol,
                side: 'BUY',
                amount: validQty,
                entryPrice: price,
                timestamp: Date.now()
            });
            
            return {
                symbol,
                side: 'BUY',
                amount: validQty,
                price,
                cost,
                orderId: `PAPER-${Date.now()}`,
                paper: true
            };
        }
        
        try {
            const order = await this.client.order({
                symbol,
                side: 'BUY',
                type: 'MARKET',
                quantity: validQty.toString()
            });
            
            const avgPrice = parseFloat(order.fills.reduce((sum, f) => 
                sum + parseFloat(f.price) * parseFloat(f.qty), 0) / 
                order.fills.reduce((sum, f) => sum + parseFloat(f.qty), 0));
            
            return {
                symbol,
                side: 'BUY',
                amount: parseFloat(order.executedQty),
                price: avgPrice,
                cost: avgPrice * parseFloat(order.executedQty),
                orderId: order.orderId,
                paper: false
            };
        } catch (error) {
            console.error(`Market buy failed for ${symbol}:`, error.message);
            throw error;
        }
    }
    
    /**
     * Execute a market sell order
     */
    async marketSell(symbol, quantity) {
        const price = await this.getPrice(symbol);
        
        // Get valid quantity for this symbol
        const validQty = await this.getValidQuantity(symbol, quantity);
        if (validQty <= 0) {
            throw new Error(`Quantity ${quantity} too small for ${symbol} (min lot size not met)`);
        }
        
        if (this.paperMode) {
            // Find and remove position
            const posIdx = this.paperPositions.findIndex(
                p => p.symbol === symbol && p.side === 'BUY'
            );
            
            if (posIdx === -1) {
                throw new Error('No position to sell');
            }
            
            const position = this.paperPositions[posIdx];
            const revenue = validQty * price;
            const pnl = revenue - (validQty * position.entryPrice);
            
            this.paperBalance += revenue;
            this.paperPositions.splice(posIdx, 1);
            
            return {
                symbol,
                side: 'SELL',
                amount: validQty,
                price,
                revenue,
                pnl,
                orderId: `PAPER-${Date.now()}`,
                paper: true
            };
        }
        
        try {
            const order = await this.client.order({
                symbol,
                side: 'SELL',
                type: 'MARKET',
                quantity: validQty.toString()
            });
            
            const avgPrice = parseFloat(order.fills.reduce((sum, f) => 
                sum + parseFloat(f.price) * parseFloat(f.qty), 0) / 
                order.fills.reduce((sum, f) => sum + parseFloat(f.qty), 0));
            
            return {
                symbol,
                side: 'SELL',
                amount: parseFloat(order.executedQty),
                price: avgPrice,
                revenue: avgPrice * parseFloat(order.executedQty),
                orderId: order.orderId,
                paper: false
            };
        } catch (error) {
            console.error(`Market sell failed for ${symbol}:`, error.message);
            throw error;
        }
    }
    
    /**
     * Get open orders
     */
    async getOpenOrders(symbol = null) {
        if (this.paperMode) {
            return this.paperPositions.filter(p => !symbol || p.symbol === symbol);
        }
        
        try {
            const params = symbol ? { symbol } : {};
            return await this.client.openOrders(params);
        } catch (error) {
            console.error('Failed to get open orders:', error.message);
            return [];
        }
    }
    
    /**
     * Get paper positions (for tracking)
     */
    getPaperPositions() {
        return this.paperPositions;
    }
    
    /**
     * Update paper position prices
     */
    async updatePaperPositionPrices() {
        if (!this.paperMode || this.paperPositions.length === 0) return;
        
        const symbols = [...new Set(this.paperPositions.map(p => p.symbol))];
        const prices = await this.getPrices(symbols);
        
        this.paperPositions.forEach(pos => {
            pos.currentPrice = prices[pos.symbol];
            pos.pnl = (pos.currentPrice - pos.entryPrice) * pos.amount;
            pos.pnlPercent = ((pos.currentPrice - pos.entryPrice) / pos.entryPrice) * 100;
        });
    }
    
    /**
     * Create a market order (used by arbitrage scanner)
     * @param {string} symbol - Trading pair (e.g., 'BTCUSDT')
     * @param {string} side - 'buy' or 'sell'
     * @param {number} quantity - Amount to trade
     */
    async createMarketOrder(symbol, side, quantity) {
        if (side.toLowerCase() === 'buy') {
            return await this.marketBuy(symbol, quantity);
        } else {
            return await this.marketSell(symbol, quantity);
        }
    }
    
    /**
     * Get real-time order book for a symbol
     */
    async getOrderBook(symbol, limit = 5) {
        try {
            const book = await this.client.book({ symbol, limit });
            return {
                bids: book.bids.map(b => ({ price: parseFloat(b.price), qty: parseFloat(b.quantity) })),
                asks: book.asks.map(a => ({ price: parseFloat(a.price), qty: parseFloat(a.quantity) }))
            };
        } catch (error) {
            console.error(`Failed to get order book for ${symbol}:`, error.message);
            return { bids: [], asks: [] };
        }
    }
}

module.exports = Exchange;

File: bot/exchange.py
"""
Binance Exchange API Wrapper
Handles all communication with Binance exchange
"""
import ccxt
import logging
from typing import Dict, List, Optional, Any
from datetime import datetime
import pandas as pd
import config

logger = logging.getLogger(__name__)


class BinanceExchange:
    """Wrapper for Binance exchange operations"""
    
    def __init__(self, paper_mode: bool = True):
        """
        Initialize Binance exchange connection
        
        Args:
            paper_mode: If True, use sandbox/simulation mode
        """
        self.paper_mode = paper_mode
        self.exchange = None
        self.paper_balance = {"USDT": config.STARTING_CAPITAL}
        self.paper_positions = {}
        self._connect()
        
    def _connect(self):
        """Establish connection to Binance"""
        try:
            self.exchange = ccxt.binance({
                'apiKey': config.BINANCE_API_KEY,
                'secret': config.BINANCE_SECRET_KEY,
                'sandbox': self.paper_mode,
                'enableRateLimit': True,
                'options': {
                    'defaultType': 'spot',
                    'adjustForTimeDifference': True,
                }
            })
            
            if not self.paper_mode and config.BINANCE_API_KEY:
                # Test connection with live API
                self.exchange.load_markets()
                logger.info("Connected to Binance (LIVE MODE)")
            else:
                self.exchange.load_markets()
                logger.info("Connected to Binance (PAPER MODE)")
                
        except Exception as e:
            logger.error(f"Failed to connect to Binance: {e}")
            raise
    
    def get_balance(self, currency: str = "USDT") -> float:
        """
        Get available balance for a currency
        
        Args:
            currency: Currency symbol (e.g., "USDT", "BTC")
            
        Returns:
            Available balance
        """
        if self.paper_mode:
            return self.paper_balance.get(currency, 0.0)
            
        try:
            balance = self.exchange.fetch_balance()
            return float(balance.get(currency, {}).get('free', 0))
        except Exception as e:
            logger.error(f"Error fetching balance: {e}")
            return 0.0
    
    def get_total_portfolio_value(self) -> float:
        """Get total portfolio value in USDT"""
        if self.paper_mode:
            total = self.paper_balance.get("USDT", 0)
            for symbol, position in self.paper_positions.items():
                current_price = self.get_current_price(symbol)
                total += position['amount'] * current_price
            return total
            
        try:
            balance = self.exchange.fetch_balance()
            total = 0.0
            for currency, amounts in balance.get('total', {}).items():
                if amounts > 0:
                    if currency == 'USDT':
                        total += amounts
                    else:
                        try:
                            ticker = self.exchange.fetch_ticker(f"{currency}/USDT")
                            total += amounts * ticker['last']
                        except:
                            pass
            return total
        except Exception as e:
            logger.error(f"Error fetching portfolio value: {e}")
            return 0.0
    
    def get_current_price(self, symbol: str) -> float:
        """
        Get current price for a trading pair
        
        Args:
            symbol: Trading pair (e.g., "BTC/USDT")
            
        Returns:
            Current price
        """
        try:
            ticker = self.exchange.fetch_ticker(symbol)
            return float(ticker['last'])
        except Exception as e:
            logger.error(f"Error fetching price for {symbol}: {e}")
            return 0.0
    
    def get_ohlcv(self, symbol: str, timeframe: str = "15m", limit: int = 100) -> pd.DataFrame:
        """
        Get OHLCV candle data
        
        Args:
            symbol: Trading pair
            timeframe: Candle timeframe (1m, 5m, 15m, 1h, etc.)
            limit: Number of candles to fetch
            
        Returns:
            DataFrame with OHLCV data
        """
        try:
            ohlcv = self.exchange.fetch_ohlcv(symbol, timeframe, limit=limit)
            df = pd.DataFrame(ohlcv, columns=['timestamp', 'open', 'high', 'low', 'close', 'volume'])
            df['timestamp'] = pd.to_datetime(df['timestamp'], unit='ms')
            df.set_index('timestamp', inplace=True)
            return df
        except Exception as e:
            logger.error(f"Error fetching OHLCV for {symbol}: {e}")
            return pd.DataFrame()
    
    def place_market_buy(self, symbol: str, amount_usdt: float) -> Optional[Dict]:
        """
        Place a market buy order
        
        Args:
            symbol: Trading pair
            amount_usdt: Amount in USDT to spend
            
        Returns:
            Order info dict or None if failed
        """
        try:
            current_price = self.get_current_price(symbol)
            if current_price <= 0:
                logger.error(f"Invalid price for {symbol}")
                return None
                
            # Calculate quantity
            quantity = amount_usdt / current_price
            
            # Get market precision
            market = self.exchange.market(symbol)
            quantity = self.exchange.amount_to_precision(symbol, quantity)
            quantity = float(quantity)
            
            if self.paper_mode:
                # Simulate order
                order = {
                    'id': f"paper_{datetime.now().timestamp()}",
                    'symbol': symbol,
                    'side': 'buy',
                    'type': 'market',
                    'price': current_price,
                    'amount': quantity,
                    'cost': amount_usdt,
                    'status': 'closed',
                    'timestamp': datetime.now().isoformat()
                }
                
                # Update paper balance
                self.paper_balance['USDT'] = self.paper_balance.get('USDT', 0) - amount_usdt
                base_currency = symbol.split('/')[0]
                self.paper_balance[base_currency] = self.paper_balance.get(base_currency, 0) + quantity
                
                # Track position
                self.paper_positions[symbol] = {
                    'amount': quantity,
                    'entry_price': current_price,
                    'entry_time': datetime.now()
                }
                
                logger.info(f"PAPER BUY: {quantity} {symbol} @ {current_price}")
                return order
            else:
                # Place real order
                order = self.exchange.create_market_buy_order(symbol, quantity)
                logger.info(f"LIVE BUY: {quantity} {symbol} @ {current_price}")
                return order
                
        except Exception as e:
            logger.error(f"Error placing buy order for {symbol}: {e}")
            return None
    
    def place_market_sell(self, symbol: str, quantity: float = None) -> Optional[Dict]:
        """
        Place a market sell order
        
        Args:
            symbol: Trading pair
            quantity: Amount to sell (if None, sells entire position)
            
        Returns:
            Order info dict or None if failed
        """
        try:
            current_price = self.get_current_price(symbol)
            if current_price <= 0:
                logger.error(f"Invalid price for {symbol}")
                return None
            
            base_currency = symbol.split('/')[0]
            
            if self.paper_mode:
                # Get paper position
                if quantity is None:
                    quantity = self.paper_balance.get(base_currency, 0)
                    
                if quantity <= 0:
                    logger.warning(f"No position to sell for {symbol}")
                    return None
                
                proceeds = quantity * current_price
                
                order = {
                    'id': f"paper_{datetime.now().timestamp()}",
                    'symbol': symbol,
                    'side': 'sell',
                    'type': 'market',
                    'price': current_price,
                    'amount': quantity,
                    'cost': proceeds,
                    'status': 'closed',
                    'timestamp': datetime.now().isoformat()
                }
                
                # Update paper balance
                self.paper_balance['USDT'] = self.paper_balance.get('USDT', 0) + proceeds
                self.paper_balance[base_currency] = self.paper_balance.get(base_currency, 0) - quantity
                
                # Remove position if fully sold
                if self.paper_balance[base_currency] <= 0:
                    self.paper_positions.pop(symbol, None)
                    self.paper_balance[base_currency] = 0
                
                logger.info(f"PAPER SELL: {quantity} {symbol} @ {current_price}")
                return order
            else:
                # Place real order
                if quantity is None:
                    balance = self.exchange.fetch_balance()
                    quantity = balance.get(base_currency, {}).get('free', 0)
                    
                if quantity <= 0:
                    logger.warning(f"No position to sell for {symbol}")
                    return None
                    
                quantity = self.exchange.amount_to_precision(symbol, quantity)
                quantity = float(quantity)
                
                order = self.exchange.create_market_sell_order(symbol, quantity)
                logger.info(f"LIVE SELL: {quantity} {symbol} @ {current_price}")
                return order
                
        except Exception as e:
            logger.error(f"Error placing sell order for {symbol}: {e}")
            return None
    
    def get_open_positions(self) -> Dict[str, Dict]:
        """Get all open positions"""
        if self.paper_mode:
            return self.paper_positions.copy()
            
        positions = {}
        try:
            balance = self.exchange.fetch_balance()
            for currency, amounts in balance.get('free', {}).items():
                if amounts > 0 and currency != 'USDT':
                    symbol = f"{currency}/USDT"
                    try:
                        current_price = self.get_current_price(symbol)
                        positions[symbol] = {
                            'amount': amounts,
                            'current_price': current_price,
                            'value': amounts * current_price
                        }
                    except:
                        pass
        except Exception as e:
            logger.error(f"Error fetching positions: {e}")
            
        return positions
    
    def get_order_book(self, symbol: str, limit: int = 10) -> Dict:
        """Get order book for a symbol"""
        try:
            return self.exchange.fetch_order_book(symbol, limit)
        except Exception as e:
            logger.error(f"Error fetching order book for {symbol}: {e}")
            return {'bids': [], 'asks': []}
